File: node/src/trufflehog.ts
/**
 * Thin wrapper over the `trufflehog` binary — find + parse to Finding[].
 *
 * vtx-recon does the *intelligence* (capability ladders + impact tiering +
 * evidence bundles); TruffleHog does the *finding* and the live/dead verify.
 * This module shells out to `trufflehog --json` and parses the NDJSON stream
 * into {@link Finding} objects.
 *
 * Importing this module is side-effect free: the binary is *not* invoked or
 * even located at import time. Detection and execution happen only when the
 * functions here are called. Tests must mock these functions / `node:child_process`
 * and never run the real binary.
 *
 * Confirmed TruffleHog JSON fields used: `DetectorName`, `Verified`, `Raw`,
 * `Redacted`, `ExtraData`. Subcommands: `git` / `github` / `filesystem` /
 * `docker` / `stdin`. Result filtering via `--results=verified,unknown,unverified`.
 */

import { spawn } from "node:child_process";
import { delimiter, join } from "node:path";
import { accessSync, constants } from "node:fs";
import { Finding } from "./models.js";

export const BINARY_NAME = "trufflehog";

const INSTALL_HINT =
  "The 'trufflehog' binary was not found on PATH.\n" +
  "vtx-recon shells out to TruffleHog for the find/verify stage.\n" +
  "Install it (https://github.com/trufflesecurity/trufflehog):\n" +
  "  brew install trufflehog\n" +
  "  # or:\n" +
  "  curl -sSfL https://raw.githubusercontent.com/trufflesecurity/" +
  "trufflehog/main/scripts/install.sh | sh -s -- -b /usr/local/bin\n" +
  "Or pass a key directly with --key / --from-trufflehog to skip the find stage.";

/** A TruffleHog invocation failed. */
export class TruffleHogError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "TruffleHogError";
  }
}

/** The trufflehog binary could not be located on PATH. */
export class TruffleHogNotFound extends TruffleHogError {
  constructor() {
    super(INSTALL_HINT);
    this.name = "TruffleHogNotFound";
  }
}

/**
 * Return the absolute path to the trufflehog binary, or throw.
 *
 * Walks `PATH` looking for an executable named `name` (with the platform's
 * executable extensions on Windows). Detection is on-demand only; never call
 * this at import time.
 */
export function findBinary(name: string = BINARY_NAME): string {
  const found = which(name);
  if (found === null) {
    throw new TruffleHogNotFound();
  }
  return found;
}

/** A cross-platform `which`: locate an executable on PATH, or null. */
function which(name: string): string | null {
  const pathEnv = process.env.PATH ?? "";
  const dirs = pathEnv.split(delimiter).filter(Boolean);
  // On Windows, try the PATHEXT extensions; elsewhere just the bare name.
  const exts =
    process.platform === "win32"
      ? (process.env.PATHEXT ?? ".EXE;.CMD;.BAT;.COM").split(";").filter(Boolean)
      : [""];

  for (const dir of dirs) {
    for (const ext of exts) {
      const candidate = join(dir, name + ext);
      try {
        accessSync(candidate, constants.X_OK);
        return candidate;
      } catch {
        // not here / not executable — keep looking.
      }
    }
  }
  return null;
}

/** A single raw TruffleHog JSON record (only the fields we read are typed). */
export interface TruffleHogRecord {
  DetectorName?: unknown;
  Verified?: unknown;
  Raw?: unknown;
  Redacted?: unknown;
  ExtraData?: unknown;
  SourceMetadata?: unknown;
  [key: string]: unknown;
}

/**
 * Map one TruffleHog JSON object to a Finding, or null if not a result.
 *
 * TruffleHog emits log lines as well as result objects on the JSON stream;
 * only objects with a `DetectorName` are results.
 */
export function parseTruffleHogRecord(record: TruffleHogRecord): Finding | null {
  const detector = record.DetectorName;
  if (!detector) {
    return null;
  }

  const extra = record.ExtraData;
  const extraData =
    extra !== null && typeof extra === "object" && !Array.isArray(extra)
      ? (extra as Record<string, unknown>)
      : {};

  const sourceMeta = record.SourceMetadata;
  const source = sourceMeta ? JSON.stringify(sourceMeta) : "";

  return new Finding({
    detectorName: String(detector),
    verified: Boolean(record.Verified ?? false),
    raw: String(record.Raw ?? ""),
    detectorRedacted: String(record.Redacted ?? ""),
    extraData,
    source,
  });
}

/**
 * Parse TruffleHog NDJSON output into Findings, skipping non-results.
 *
 * Malformed lines and non-result objects (log lines) are skipped silently so a
 * noisy stream still yields its real findings.
 */
export function parseJsonStream(lines: Iterable<string>): Finding[] {
  const findings: Finding[] = [];
  for (const raw of lines) {
    const line = raw.trim();
    if (!line) {
      continue;
    }
    let record: unknown;
    try {
      record = JSON.parse(line);
    } catch {
      continue;
    }
    if (record === null || typeof record !== "object" || Array.isArray(record)) {
      continue;
    }
    const finding = parseTruffleHogRecord(record as TruffleHogRecord);
    if (finding !== null) {
      findings.push(finding);
    }
  }
  return findings;
}

export interface RunTruffleHogOptions {
  /** Additional flags (e.g. `["--results=verified,unknown"]`). */
  extraArgs?: string[];
  /** Override the resolved binary path (used by tests). */
  binary?: string;
  /** Milliseconds before the scan is aborted. Default 300_000 (300s). */
  timeoutMs?: number;
}

/**
 * Run `trufflehog <subcommand> <target> --json` and parse the results.
 *
 * @param subcommand one of git / github / filesystem / docker / stdin.
 * @param target the scan target (repo URL, path, image, ...).
 *
 * @throws {@link TruffleHogNotFound} the binary is not installed.
 * @throws {@link TruffleHogError} the scan failed to run.
 */
export async function runTruffleHog(
  subcommand: string,
  target: string,
  options: RunTruffleHogOptions = {},
): Promise<Finding[]> {
  const exe = options.binary ?? findBinary();
  const timeoutMs = options.timeoutMs ?? 300_000;
  const args = [subcommand, target, "--json", ...(options.extraArgs ?? [])];

  const { stdout, stderr, code, error } = await spawnCapture(exe, args, timeoutMs);

  if (error) {
    if ((error as NodeJS.ErrnoException).code === "ENOENT") {
      // Binary vanished between which() and spawn.
      throw new TruffleHogNotFound();
    }
    if ((error as { vtxTimeout?: boolean }).vtxTimeout) {
      throw new TruffleHogError(`trufflehog timed out after ${timeoutMs / 1000}s`);
    }
    throw new TruffleHogError(`trufflehog failed to run: ${error.message}`);
  }

  // TruffleHog exits non-zero when it finds verified secrets, so a non-zero
  // code is not on its own an error; only treat it as failure if there is no
  // parseable JSON output at all.
  const findings = parseJsonStream(stdout.split(/\r?\n/));
  if (findings.length === 0 && code !== 0 && code !== 183) {
    throw new TruffleHogError(
      `trufflehog exited ${code} with no findings: ${stderr.trim() || "(no stderr)"}`,
    );
  }
  return findings;
}

interface SpawnCapture {
  stdout: string;
  stderr: string;
  code: number | null;
  error: Error | null;
}

/** Spawn a process, capture stdout/stderr, and enforce a timeout. */
function spawnCapture(exe: string, args: string[], timeoutMs: number): Promise<SpawnCapture> {
  return new Promise<SpawnCapture>((resolve) => {
    const child = spawn(exe, args, { stdio: ["ignore", "pipe", "pipe"] });
    let stdout = "";
    let stderr = "";
    let settled = false;

    const timer = setTimeout(() => {
      if (settled) return;
      const timeoutError = Object.assign(new Error("timed out"), { vtxTimeout: true });
      child.kill("SIGKILL");
      finish({ stdout, stderr, code: null, error: timeoutError });
    }, timeoutMs);

    const finish = (result: SpawnCapture): void => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      resolve(result);
    };

    child.stdout?.on("data", (chunk: Buffer) => {
      stdout += chunk.toString("utf8");
    });
    child.stderr?.on("data", (chunk: Buffer) => {
      stderr += chunk.toString("utf8");
    });
    child.on("error", (err: Error) => finish({ stdout, stderr, code: null, error: err }));
    child.on("close", (code: number | null) => finish({ stdout, stderr, code, error: null }));
  });
}

File: python/src/vtx_recon/trufflehog.py
"""Thin wrapper over the ``trufflehog`` binary — find + parse to Finding[].

vtx-recon does the *intelligence* (capability ladders + impact tiering +
evidence bundles); TruffleHog does the *finding* and the live/dead verify.
This module shells out to ``trufflehog --json`` and parses the NDJSON stream
into :class:`~vtx_recon.models.Finding` objects.

Importing this module is side-effect free: the binary is *not* invoked at
import time. Detection and execution happen only when the functions here are
called. Tests must mock these functions and never run the real binary.

Confirmed TruffleHog JSON fields used: ``DetectorName``, ``Verified``,
``Raw``, ``Redacted``, ``ExtraData``. Subcommands: ``git`` / ``github`` /
``filesystem`` / ``docker`` / ``stdin``. Result filtering via
``--results=verified,unknown,unverified``.
"""

from __future__ import annotations

import json
import shutil
import subprocess
from collections.abc import Iterable, Iterator

from .models import Finding

__all__ = [
    "BINARY_NAME",
    "TruffleHogError",
    "TruffleHogNotFound",
    "find_binary",
    "parse_json_stream",
    "parse_trufflehog_record",
    "run_trufflehog",
]

BINARY_NAME = "trufflehog"

_INSTALL_HINT = (
    "The 'trufflehog' binary was not found on PATH.\n"
    "vtx-recon shells out to TruffleHog for the find/verify stage.\n"
    "Install it (https://github.com/trufflesecurity/trufflehog):\n"
    "  brew install trufflehog\n"
    "  # or:\n"
    "  curl -sSfL https://raw.githubusercontent.com/trufflesecurity/"
    "trufflehog/main/scripts/install.sh | sh -s -- -b /usr/local/bin\n"
    "Or pass a key directly with --key / --from-trufflehog to skip the find "
    "stage."
)


class TruffleHogError(RuntimeError):
    """A TruffleHog invocation failed."""


class TruffleHogNotFound(TruffleHogError):
    """The trufflehog binary could not be located on PATH."""

    def __init__(self) -> None:
        super().__init__(_INSTALL_HINT)


def find_binary(name: str = BINARY_NAME) -> str:
    """Return the absolute path to the trufflehog binary, or raise.

    Detection is on-demand only; never call this at import time.
    """
    path = shutil.which(name)
    if path is None:
        raise TruffleHogNotFound()
    return path


def parse_trufflehog_record(record: dict[str, object]) -> Finding | None:
    """Map one TruffleHog JSON object to a Finding, or None if not a result.

    TruffleHog emits log lines as well as result objects on the JSON stream;
    only objects with a ``DetectorName`` are results.
    """
    detector = record.get("DetectorName")
    if not detector:
        return None

    extra = record.get("ExtraData")
    extra_data: dict[str, object] = extra if isinstance(extra, dict) else {}

    source_meta = record.get("SourceMetadata")
    source = json.dumps(source_meta, separators=(",", ":")) if source_meta else ""

    return Finding(
        detector_name=str(detector),
        verified=bool(record.get("Verified", False)),
        raw=str(record.get("Raw", "")),
        detector_redacted=str(record.get("Redacted", "")),
        extra_data=extra_data,
        source=source,
    )


def parse_json_stream(lines: Iterable[str]) -> Iterator[Finding]:
    """Parse TruffleHog NDJSON output into Findings, skipping non-results.

    Malformed lines and non-result objects (log lines) are skipped silently
    so a noisy stream still yields its real findings.
    """
    for line in lines:
        line = line.strip()
        if not line:
            continue
        try:
            record = json.loads(line)
        except json.JSONDecodeError:
            continue
        if not isinstance(record, dict):
            continue
        finding = parse_trufflehog_record(record)
        if finding is not None:
            yield finding


def run_trufflehog(
    subcommand: str,
    target: str,
    *,
    extra_args: list[str] | None = None,
    binary: str | None = None,
    timeout: float = 300.0,
) -> list[Finding]:
    """Run ``trufflehog <subcommand> <target> --json`` and parse the results.

    Args:
        subcommand: one of git / github / filesystem / docker / stdin.
        target: the scan target (repo URL, path, image, ...).
        extra_args: additional flags (e.g. ``["--results=verified,unknown"]``).
        binary: override the resolved binary path (used by tests).
        timeout: seconds before the scan is aborted.

    Raises:
        TruffleHogNotFound: the binary is not installed.
        TruffleHogError: the scan failed to run.
    """
    exe = binary or find_binary()
    cmd = [exe, subcommand, target, "--json", *(extra_args or [])]
    try:
        proc = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            timeout=timeout,
            check=False,
        )
    except FileNotFoundError as exc:  # binary vanished between which() and run
        raise TruffleHogNotFound() from exc
    except subprocess.TimeoutExpired as exc:
        raise TruffleHogError(f"trufflehog timed out after {timeout}s") from exc

    # TruffleHog exits non-zero when it finds verified secrets, so a non-zero
    # code is not on its own an error; only treat it as failure if there is
    # no parseable JSON output at all.
    findings = list(parse_json_stream(proc.stdout.splitlines()))
    if not findings and proc.returncode not in (0, 183):
        raise TruffleHogError(
            f"trufflehog exited {proc.returncode} with no findings: "
            f"{proc.stderr.strip() or '(no stderr)'}"
        )
    return findings
